core/domain/topic_services.py

```diff
@@ -1763,3 +1792,54 @@ def get_chapter_counts_in_topic_summaries(
         })
 
     return topic_chapter_counts_dict
+
+
+def get_all_published_story_exploration_ids(
+    topic_id: Optional[str] = None
+) -> List[str]:
+    """Returns a list of each exploration id linked to each published story
+    chapter that belongs to topic(s).
+
+    Args:
+        topic_id: str|None. The id of a topic. If topic_id is provided, the
+            result includes exploration ids only from the corresponding
+            topic. If no topic_id is provided, the result includes
+            exploration ids in all topics.
+
+    Returns:
+        list(str). A list of all exploration ids linked to the topic(s)'
+        published stories' chapters.
+    """
+    fetched_topic_summaries = (
+        [topic_fetchers.get_topic_summary_by_id(topic_id)] if topic_id
+        else topic_fetchers.get_all_topic_summaries())
+
+    # Keep all topic summaries with a mapping. For those without a mapping,
+    # record their ids, fetch their corresponding topics with them, and then
+    # use the topic to compute their summary. Add each new topic summary to
+    # the list of summaries that initially had a mapping.
+    topic_summaries_with_persisted_mapping = []
+    ids_of_summaries_without_persisted_mapping: List[str] = []
+    for summary in fetched_topic_summaries:
+        if summary.published_story_exploration_mapping is None:
+            ids_of_summaries_without_persisted_mapping.append(summary.id)
+        else:
+            topic_summaries_with_persisted_mapping.append(summary)
+    topic_summaries_with_recently_computed_mapping = []
+    if len(ids_of_summaries_without_persisted_mapping) > 0:
+        topics = topic_fetchers.get_topics_by_ids(
+            ids_of_summaries_without_persisted_mapping)
+        for topic in topics:
+            assert topic is not None
+            topic_summaries_with_recently_computed_mapping.append(
+                compute_summary_of_topic(topic))
+    topic_summaries = (
+        topic_summaries_with_persisted_mapping +
+        topic_summaries_with_recently_computed_mapping)
+
+    exp_ids = itertools.chain.from_iterable(
+        itertools.chain.from_iterable(
+            summary.published_story_exploration_mapping.values())
+        for summary in topic_summaries)
+
+    return list(set(exp_ids))
```

Comment:
If all you want from this method is to get the exp_ids, rather than saving the regenerated summaries, then why bother computing the summaries for the topics without a persisted mapping (rather than just getting the data directly from the TopicModel)?

Reply:
`TopicModel` doesn't contain any information about its stories' `exp_ids`. Are you suggesting that the app fetch the relevant story ids from `TopicModel`, and then fetch the linked `exp_ids` from the stories corresponding to the story ids? If this is the case, then I suppose that I initially misjudged recomputing the summary as being quicker than getting that data directly from the datastore. Both ways require fetching the relevant stories and exp_ids, so I now think this suggestion would be much quicker and simpler.

Reply:
Yeah, I would do that (only if the data doesn't exist in the topic summary). Otherwise it seems like a lot of extra work (at least conceptually) to generate the summary and not store it. (And storing it probably isn't a good idea here since it introduces a PUT side-effect to this function.)

Reply:
Done.